config/decoder.go

```diff
@@ -369,18 +315,35 @@ func (d *decoder) unmarshal(name string, value reflect.Value, def string) error
 		d.m[val] = struct{}{}
 	}
 
-	switch getBucket(value.Type()) {
-	case bucketPrimitive:
-		return d.scalar(name, value, def)
-	case bucketObject:
+	return nil
+}
+
+// Dispatch un-marshalling functions based on the value type.
+func (d *decoder) unmarshal(name string, value reflect.Value, def string) error {
+	if err := d.checkCycles(value); err != nil {
+		return err
+	}
+
+	switch value.Kind() {
+	case reflect.Invalid:
+		return fmt.Errorf("invalid value type for key %s", name)
+	case reflect.Ptr:
+		return d.pointer(name, value, def)
+	case reflect.Struct:
 		return d.object(name, value)
-	case bucketArray:
+	case reflect.Array:
 		return d.array(name, value)
-	case bucketSlice:
+	case reflect.Slice:
 		return d.sequence(name, value)
-	case bucketMap:
+	case reflect.Map:
+		return d.mapping(name, value, def)
+	case reflect.Interface:
 		return d.mapping(name, value, def)
+	case reflect.Chan:
+		return nil
```

Comment:
We iterate over fields, so if a struct has a channel, we want to simply skip it.

Actually we should check that this channel also implements TextUnmarshaller to possibly construct it, but this is a subject of other change. Example is zap.Config, some of it's fields are actually functions that implement unmarshaller to read configs.